Resources/Storage/S3Storage.py

```diff
@@ -0,0 +1,748 @@
+    :returns: Successful dict: {path : size}
+              Failed dict: {path : errorMessage}
+              S_ERROR in case of argument problems
+    """
+
+    log = LOG.getSubLogger('getFile')
+
+    failed = {}
+    successful = {}
+
+    res = self.s3GWClient.createPresignedUrl(self.name, 'get_object', urls)
+    if not res['OK']:
+      return res
+
+    failed.update(res['Value']['Failed'])
+
+    presignedURLs = res['Value']['Successful']
+
+    for src_url, presignedURL in presignedURLs.items():
+      try:
+        fileName = os.path.basename(src_url)
+        dest_file = os.path.join(localPath if localPath else os.getcwd(), fileName)
+        log.debug("Trying to download %s to %s" % (src_url, dest_file))
+
+        # Stream download to save memory
+        # https://requests.readthedocs.io/en/latest/user/advanced/#body-content-workflow
+        with requests.get(presignedURL, stream=True) as r:
+          r.raise_for_status()
+          with open(dest_file, 'wb') as f:
+            for chunk in r.iter_content():
+              if chunk:  # filter out keep-alive new chuncks
+                f.write(chunk)
+
+        successful[src_url] = os.path.getsize(dest_file)
+      except Exception as exp:
+        failed[src_url] = repr(exp)
+
+    return S_OK({'Failed': failed, 'Successful': successful})
+
+  def putFile(self, urls, sourceSize=0):
+    """ Upload a local file.
+        :warning: no 3rd party copy possible
+
+        :param urls: dictionary { urls : localFile }
+        :param sourceSize: size of the file in byte. Mandatory for third party copy (WHY ???)
+                             Also, this parameter makes it essentially a non bulk operation for
+                             third party copy, unless all files have the same size...
+        :returns: Successful dict: { path : size }
+                  Failed dict: { path : error message }
+                  S_ERROR in case of argument problems
+    """
+    if self.directAccess:
+      return self._direct_putFile(urls, sourceSize=sourceSize)
+    return self._presigned_putFile(urls, sourceSize=sourceSize)
+
+  @_extractKeyFromS3Path
+  def _direct_putFile(self, urls, sourceSize=0):
+    """ Upload a local file.
+        :warning: no 3rd party copy possible
+
+        :param urls: dictionary { urls : localFile }
+        :param sourceSize: size of the file in byte. Mandatory for third party copy (WHY ???)
+                             Also, this parameter makes it essentially a non bulk operation for
+                             third party copy, unless all files have the same size...
+        :returns: Successful dict: { path : size }
+                  Failed dict: { path : error message }
+                  S_ERROR in case of argument problems
+    """
+
+    log = LOG.getSubLogger('putFile')
+
+    # the @_extractKeyFromS3Path transformed URL into keys
+    keys = urls
+
+    failed = {}
+    successful = {}
+
+    for dest_key, src_file in keys.iteritems():
```

Comment:
Why?

Reply:
python 3 :cry:

Reply:
In the hope that one day DIRAC supports Python 3. Every line that is added and later needs to be changed is another 1% of a bug when the code is ported.

I'll justify why at in a BiLD meeting a some point (it could even be this Thursday?) but every modern recommendation for large projects is to make the same code base work with both versions. For example:

* https://portingguide.readthedocs.io/en/latest/index.html
* https://gregoryszorc.com/blog/2020/01/13/mercurial%27s-journey-to-and-reflections-on-python-3/

Reply:
Btw, don't forget the "how-to" list we discussed at the last bild :)

Reply:
For `items` vs `iteritems` please see discussion in https://github.com/DIRACGrid/DIRAC/pull/4243#issuecomment-523920369

Reply:
So what with `items` vs `iteritems` ? I know we argued a lot about that, and concluded that it was better to make explicit what we wanted to do, in the hope that `2to3` would happily run on it. But now that we know that `2to3` is not going to make it, what do we do ? I still think it is nice to make it explicit where it needs to actually be a copy, but is there a way to do it avoiding the ugly double copy of `list(dict.items())` ?

Reply:
made them all `items`